### ISSUE TITLE
Dynamic BigQuery table name generation?

### ISSUE THREAD

**Issue by jonopwell** [spotify/scio#434]
This may be more of a lack in my Scala knowledge rather than scio, but we are wondering how we can pick up the event that is about to be saved to BigQuery and dynamically generate its table name by calling an arbitrary method.

For example (sorry about the bad formatting),

`def getTableId(r: (OurDataType,TableRow)):String = {
      r._1.toString()  // generate tableId here based on elements in OurDataType
}
  sideouts(sideOutputKey)  .map(x => OurDataTypeConverter.format(x._1.asInstanceOf[OurDataType]))  .setCoder(TableRowJsonCoder.of)  .saveAsBigQuery(getTableId(_), OurDataTypeTableSchemas.schema,
        WriteDisposition.WRITE_APPEND, CreateDisposition.CREATE_IF_NEEDED)
`

Gets compile error:

`Error:(60, 8) overloaded method value saveAsBigQuery with alternatives:
  (tableSpec: String,schema: com.google.api.services.bigquery.model.TableSchema,writeDisposition: com.google.cloud.dataflow.sdk.io.BigQueryIO.Write.WriteDisposition,createDisposition: com.google.cloud.dataflow.sdk.io.BigQueryIO.Write.CreateDisposition)(implicit ev: <:<[com.spotify.scio.bigquery.TableRow,com.google.api.services.bigquery.model.TableRow])scala.concurrent.Future[com.spotify.scio.io.Tap[com.google.api.services.bigquery.model.TableRow]] <and>
  (table: com.google.api.services.bigquery.model.TableReference,schema: com.google.api.services.bigquery.model.TableSchema,writeDisposition: com.google.cloud.dataflow.sdk.io.BigQueryIO.Write.WriteDisposition,createDisposition: com.google.cloud.dataflow.sdk.io.BigQueryIO.Write.CreateDisposition)(implicit ev: <:<[com.spotify.scio.bigquery.TableRow,com.google.api.services.bigquery.model.TableRow])scala.concurrent.Future[com.spotify.scio.io.Tap[com.google.api.services.bigquery.model.TableRow]]
 cannot be applied to (((com.dexcom.stream.processing.OurDataType, com.google.api.services.bigquery.model.TableRow)) => String, com.google.api.services.bigquery.model.TableSchema, com.google.cloud.dataflow.sdk.io.BigQueryIO.Write.WriteDisposition, com.google.cloud.dataflow.sdk.io.BigQueryIO.Write.CreateDisposition)
      .saveAsBigQuery(getTableId(_), OurDataTypeTableSchemas.schema,`

Bottom line question: how can we get access to the event in the implicit so we can dynamically generate the tableId in the saveAsBigQuery/getTableId() call? It appears that the "<:<" operator is just ensuring correct type, but appears to complicate access to the actual event in this situation.

Thanks for any help!

~Jon

P.S. Our architecture is hardening, and I plan to add to the AsPoweredBy once we successfully progress further.

**Comment by nevillelyh** [comment]
IIRC SDK's BigQueryIO has something like that called toTableReference. I can look into it.

Most of us will be out for about a week so we'll pick it up later.

**Comment by ravwojdyla** [comment]
I'm afraid you are not going to be able to use Scio API. `saveAsBigQuery` is executed before job is submitted, before any events are available. You could instead use `saveAsCustomOutput` and provide it with your own `BigQueryIO.Write`, where you provide SerializableFunction to shard BigQuery output. Please check this [doc](https://cloud.google.com/dataflow/java-sdk/JavaDoc/com/google/cloud/dataflow/sdk/io/BigQueryIO#sharding-bigquery-output-tables). Does this answer your question?

**Comment by jonopwell** [comment]
@ravwojdyla : perfect. That does exactly what we need. Thanks! Appreciate the rapid response.

P.S. We are asking our management about getting on your "powered-by" section on the Wiki.

**Comment by ravwojdyla** [comment]
@jonopwell thanks - would be awesome to have you in the powered-by!

**Comment by sfines** [comment]
Can this technique be applied to the GCS writes with saveAsTextfile ? and If so, is there an example on how to use a SerializedFunction inside of SCIO?

**Comment by nevillelyh** [comment]
Don't think the file IO has such feature. You'll have to do it at a much lower level, e.g. using a `GcsUtil` in each `.map` transform to create files and write to them.
